pypicloud/cache/dynamo.py

```diff
@@ -22,6 +25,91 @@
 LOG = logging.getLogger(__name__)
 
 
+def _clear_summary(summary):
+
+    """ Clear out the field of a PackageSummary object """
+    summary.stable = None
+    summary.unstable = '0'
+    summary.last_modified = datetime.fromtimestamp(0).replace(tzinfo=UTC)
+    return summary
+
+
+CacheUpdates = namedtuple(
+    'CacheUpdates',
+    [
+        'new_packages',
+        'seen_packages',
+        'stale_packages',
+        'new_summaries',
+        'seen_summaries',
+        'stale_summaries',
+    ],
+)
+
+
+def calculate_cache_updates(cache_packages, cache_summaries, storage_packages, summary_factory):
+
+    """ Calculates what changes need to happen in the cache to update it. """
+    # Clear out non-key fields of the cached summaries to purge stable and
+    # unstable fields.
+    cached_summ_by_name = dict(
+        (summ.name, _clear_summary(summ),)
+        for summ in cache_summaries
+    )
+    cached_pkg_by_filename = dict(
+        (pkg.filename, pkg,)
+        for pkg in cache_packages
+    )
+
+    new_packages = set()
+    seen_packages = set()
+    new_summaries_by_name = dict()
+    seen_summaries = set()
+    for pkg in storage_packages:
+        try:
+            cached_pkg = cached_pkg_by_filename[pkg.filename]
+        except KeyError:
+            new_packages.add(pkg)
+        else:
+            seen_packages.add(cached_pkg)
+
+            # Skip over duplicate files (same filename, different paths).
+            # XXX Is this acceptable? How should I decide this?
+            if cached_pkg.last_modified > pkg.last_modified:
```

Comment:
I would be inclined to say that if the filename exists in the cache then the package can be ignored entirely. Do you have cases where the cache contains a filename but needs to update the other data?